### PR TITLE
PP-339: additional QA test for node ramp down

### PR DESCRIPTION
**Pull request by anamikau**
#### Issue-ID

* **[PP-339](https://pbspro.atlassian.net/browse/PP-339)**

#### Problem description
* Added few more regressions for node ramp down feature as part of QA verifications

#### Cause / Analysis
* Added few more regressions for node ramp down feature as part of QA verifications

#### Solution description
* Please describe your changes in detail for reviewers

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/tests/functional/pbs_node_rampdown.py [hunk 1]
```diff
@@ -6362,3 +6415,774 @@ def test_release_nodes_shared_server_restart_immed(self):
...
+        ret = self.server.du.run_cmd(self.server.hostname, cmd)
+        self.assertEqual(ret['rc'], 0)
+
+        # Submit another job and make sure it is
+        # picked up by hostC
+        j = Job(TEST_USER,
+                {'Resource_List.select': "1:host=" + self.hostC})
+        jid2 = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': "R",
+                                 "exec_host": "centos3/1"}, id=jid2)
```
Comment (bhroam) [inline test/tests/functional/pbs_node_rampdown.py]: I'm assuming that centos3 is a hostname?  Putting static host names in PTL tests is bad since the test will fail on any other host.  Maybe use self.momX.short_hostname?

File: test/tests/functional/pbs_node_rampdown.py [hunk 2]
```diff
@@ -6362,3 +6415,774 @@ def test_release_nodes_shared_server_restart_immed(self):
...
+                                 'Resource_List.select': newsel,
+                                 'Resource_List.place': self.job1_place,
+                                 'Resource_List.nodect': 3,
+                                 'schedselect': newsel,
+                                 'exec_host': self.job1_exec_host,
+                                 'exec_vnode': new_exec_vnode}, id=jid)
+
+        # Release the job and make sure it is running
+        self.server.sigjob(jid, 'resume')
+        self.server.expect(JOB, {'job_state': 'R',
```
Comment (bhroam) [inline test/tests/functional/pbs_node_rampdown.py]: I don't really see resuming the job as part of this test.  You are testing to see that you can't release nodes from a suspended job.  This has been done.  Resuming the job is just testing that you can resume jobs.

File: test/tests/functional/pbs_node_rampdown.py [hunk 3]
```diff
@@ -6362,3 +6415,774 @@ def test_release_nodes_shared_server_restart_immed(self):
...
+
+        # Verify that 3 processes running on hostB
+        process = 0
+        self.server.pu.get_proc_info(
+            self.momB.hostname, ".*fib.*", None, regexp=True)
+        if (self.server.pu.processes is not None):
+            for key in self.server.pu.processes:
+                if ("fib" in key):
+                    process = len(self.server.pu.processes[key])
+                    print "length of the process is %d " % (process,)
```
Comment (bhroam) [inline test/tests/functional/pbs_node_rampdown.py]: Don't use print, use self.logger.  I don't find this log message useful as it is.  I'd either add " (expected 3)" to the end or remove the message.  Same for the other use of process.

File: test/tests/functional/pbs_node_rampdown.py [hunk 4]
```diff
@@ -6362,3 +6415,774 @@ def test_release_nodes_shared_server_restart_immed(self):
...
+            max_attempts=18, interval=2, regexp=True)
+
+        self.momA.log_match(
+            "Job;%s;%s.+cput=.+mem.+" % (jid2, self.hostB),
+            max_attempts=8, regexp=True,
+            existence=False)
+
+        self.momA.log_match(
+            "Job;%s;%s.+cput=.+mem.+" % (jid3, self.hostB),
+            max_attempts=5, regexp=True,
```
Comment (bhroam) [inline test/tests/functional/pbs_node_rampdown.py]: Why the difference in max_attempts (8 vs 5) ?

File: test/tests/functional/pbs_node_rampdown.py [hunk 5]
```diff
@@ -6362,3 +6415,774 @@ def test_release_nodes_shared_server_restart_immed(self):
...
+                                 "Resource_List.nodect": 3}, id=jid2)
+        self.server.expect(JOB, {'job_state': 'R',
+                                 "exec_host": e_host_j3,
+                                 "exec_vnode": e_vnode,
+                                 "Resource_List.nodect": 3}, id=jid3)
+
+    def test_PBS_JOBID(self):
+        """
+        Test that if jobid is not provided then it is
+        picked by env variable $PBS_JOBID
```
Comment (bhroam) [inline test/tests/functional/pbs_node_rampdown.py]: I don't see you actually call pbs_release_nodes.  Also could you change the docstring to say "if -j jobid is not provided" (add the -j)

File: test/tests/functional/pbs_node_rampdown.py [hunk 6]
```diff
@@ -6362,3 +6415,774 @@ def test_release_nodes_shared_server_restart_immed(self):
...
+            MGR_CMD_CREATE, RSC, attr, id=r, runas=ROOT_USER, logerr=False)
+
+        attr1 = {}
+        attr1['type'] = 'size'
+        attr1['flag'] = 'nh'
+        r1 = 'foo_i'
+        self.server.manager(
+            MGR_CMD_CREATE, RSC, attr1, id=r1, runas=ROOT_USER, logerr=False)
+
+        self.logger.info("test_epilogue")
```
Comment (bhroam) [inline test/tests/functional/pbs_node_rampdown.py]: Can you make this log message more explicit?  It doesn't say much.

File: test/tests/functional/pbs_node_rampdown.py [hunk 7]
```diff
@@ -6362,3 +6415,774 @@ def test_release_nodes_shared_server_restart_immed(self):
...
+        cmd = [self.pbs_release_nodes_cmd, '-j', jid, self.n7]
+        ret = self.server.du.run_cmd(self.server.hostname, cmd)
+        self.assertEqual(ret['rc'], 0)
+
+        self.momC.log_match("executed epilogue hook", max_attempts=10)
+        self.momC.log_match("DELETE_JOB2 received", max_attempts=10)
+
+        self.server.delete(jid, wait=True)
+
+        self.server.expect(JOB, {'job_state': 'F',
```
Comment (bhroam) [inline test/tests/functional/pbs_node_rampdown.py]: I don't see you enabling job history in this test.

File: test/tests/functional/pbs_node_rampdown.py [hunk 8]
```diff
@@ -6362,3 +6415,774 @@ def test_release_nodes_shared_server_restart_immed(self):
...
+                                 "resources_used.foo_i": "3kb",
+                                 "resources_used.foo_f": '0.29',
+                                 "resources_used.vmem": '29gb'}, id=jid)
+
+    @timeout(500)
+    def test_release_reservations(self):
+        """
+        Release nodes from a reservation wil throw error. However
```
Comment (bhroam) [inline test/tests/functional/pbs_node_rampdown.py]: wil=will

File: test/tests/functional/pbs_node_rampdown.py [hunk 9]
```diff
@@ -6362,3 +6415,774 @@ def test_release_nodes_shared_server_restart_immed(self):
...
+        jobs inside reservation queue works as expected.
+        """
+
+        # Create a reservation on multiple nodes
+        start = int(time.time()) + 60
+        a = {'Resource_List.select': self.job1_select,
+             'Resource_List.place': 'scatter',
+             'reserve_start': start}
```
Comment (bhroam) [inline test/tests/functional/pbs_node_rampdown.py]: Do you not have to submit a duration or an end time?

File: test/tests/functional/pbs_node_rampdown.py [hunk 10]
```diff
@@ -6362,3 +6415,774 @@ def test_release_nodes_shared_server_restart_immed(self):
...
+
+        # Submit a job inside reservations and release vnode
+        a = {'queue': rid,
+             'Resource_List.select': self.job1_select}
+        j = Job(TEST_USER, a)
+        jid = self.server.submit(j)
+
+        # Wait for the job to start
+        self.server.expect(JOB, {'job_state': 'R'},
+                           offset=60, id=jid)
```
Comment (bhroam) [inline test/tests/functional/pbs_node_rampdown.py]: Can we shorten this to 30 (and reserve_start) or will that cause problems?

File: test/tests/functional/pbs_node_rampdown.py [hunk 11]
```diff
@@ -37,6 +37,7 @@
 # trademark licensing policies.
 
 from tests.functional import *
+from ptl.lib.pbs_testlib import BatchUtils
```
[the remark this answers is not in the document]
Comment (anamikau) [reply test/tests/functional/pbs_node_rampdown.py]: Ah that was not necessary. forgot to remove it.

File: test/tests/functional/pbs_node_rampdown.py [hunk 12]
```diff
@@ -331,6 +354,10 @@ def setUp(self):
             'pbs_python -c "exec(\\\"def fib(i):\\n if i < 2:\\n  \
 return i\\n return fib(i-1) + fib(i-2)\\n\\nprint fib(50)\\\")"'
 
+        FIB400 = os.path.join(self.server.pbs_conf['PBS_EXEC'], 'bin', '') + \
+            'pbs_python -c "exec(\\\"def fib(i):\\n if i < 2:\\n  \
+return i\\n return fib(i-1) + fib(i-2)\\n\\nprint fib(400)\\\")"'
+
```
[the remark this answers is not in the document]
Comment (anamikau) [reply test/tests/functional/pbs_node_rampdown.py]: I was hitting race condition with Fib(50) hence added this.

File: test/tests/functional/pbs_node_rampdown.py [hunk 13]
```diff
@@ -521,14 +563,25 @@ def setUp(self):
...
             SLEEP_CMD + " 60"
 
+        self.job13_select = "3:ncpus=1"
+        self.script['job13'] = \
+            "#PBS -l select=" + self.job13_select + "\n" + \
+            "#PBS -l place=" + self.job1_place + "\n" + \
+            "pbsdsh -n 1 -- %s\n" % (FIB400,) + \
+            "pbsdsh -n 2 -- %s\n" % (FIB400,) + \
+            "pbsdsh -n 3 -- %s\n" % (FIB400,) + \
+            "%s\n" % (FIB50,)
```
[the remark this answers is not in the document]
Comment (anamikau) [reply test/tests/functional/pbs_node_rampdown.py]: you are right .. will fix that.

File: test/tests/functional/pbs_node_rampdown.py (hunk 1 above)
Comment (anamikau) [reply test/tests/functional/pbs_node_rampdown.py]: Oops! thanks for catching that. It was left by mistake.

File: test/tests/functional/pbs_node_rampdown.py (hunk 2 above)
Comment (anamikau) [reply test/tests/functional/pbs_node_rampdown.py]: resuming and also validating the job resources making sure nothing has changed. I would leave it like that.

File: test/tests/functional/pbs_node_rampdown.py (hunk 3 above)
Comment (anamikau) [reply test/tests/functional/pbs_node_rampdown.py]: sorry left as part of debugging. Will change.

File: test/tests/functional/pbs_node_rampdown.py (hunk 4 above)
Comment (anamikau) [reply test/tests/functional/pbs_node_rampdown.py]: can't recall. May be just want to make sure it is not there as things on VMs are slower

File: test/tests/functional/pbs_node_rampdown.py (hunk 5 above)
Comment (anamikau) [reply test/tests/functional/pbs_node_rampdown.py]: Yes it is defined in the job script. updated the docstring

File: test/tests/functional/pbs_node_rampdown.py (hunk 9 above)
Comment (anamikau) [reply test/tests/functional/pbs_node_rampdown.py]: PTL will take default.

File: test/tests/functional/pbs_node_rampdown.py (hunk 10 above)
Comment (anamikau) [reply test/tests/functional/pbs_node_rampdown.py]: let me try. Reason I kept 60 as I have noticed pbs_rsub will round off the start time and not actually now+30 causing race conditions in the test.

File: test/tests/functional/pbs_node_rampdown.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (bhroam) [reply test/tests/functional/pbs_node_rampdown.py]: Interesting, I wouldn't have thought it would end during a test.  400 definitely won't.

File: test/tests/functional/pbs_node_rampdown.py [hunk 14]
```diff
@@ -569,17 +571,17 @@ def setUp(self):
...
             "pbsdsh -n 1 -- %s\n" % (FIB400,) + \
             "pbsdsh -n 2 -- %s\n" % (FIB400,) + \
-            "pbsdsh -n 3 -- %s\n" % (FIB400,) + \
-            "%s\n" % (FIB50,)
+            "pbsdsh -n 3 -- %s\n" % (FIB400,)
 
-        self.pbs_release_nodes_cmd = os.path.join(
-            self.server.pbs_conf['PBS_EXEC'], 'bin', 'pbs_release_nodes')
+        # self.pbs_release_nodes_cmd = os.path.join(
+        #    self.server.pbs_conf['PBS_EXEC'], 'bin', 'pbs_release_nodes')
```
Comment (bhroam) [inline test/tests/functional/pbs_node_rampdown.py]: You moved this up top.  Can you remove the commented section?

**Comment by bhroam** [comment]
Codacy has some comments.  Mostly some unused variables.  Consider making those changes.

**Comment by anamikau** [comment]
@bhroam and @bayucan, addressed most of the codacy errors. Please resign off.

**Comment by bhroam** [comment]
Still looks good to me.  Thanks for making the changes.

**Comment by anamikau** [comment]
@bhroam or @PBSPro/pbspro-maintainers , rebased and squashed. Please take it forward.

**Comment by mike0042** [comment]
@anamika: The appveyor build failed because it couldn't find the wineditline package. Unfortunately, I can't restart your build as I would for Travis. There are two unused variable warnings in the codacy output that should probably be fixed. A new commit will cause the appveyor build to restart, so I suggest you address them.

File: test/tests/functional/pbs_node_rampdown.py [hunk 15]
```diff
@@ -6634,8 +6634,12 @@ def test_release_job_array(self):
                  "1:ncpus=2:mem=2097152kb"
         newsel_esc = newsel.replace("+", "\+")
         new_exec_host = self.job1_exec_host
-        new_exec_host_esc = self.job1_exec_host.replace(
-            "*", "\*").replace("[", "\[").replace("]", "\]").replace("+", "\+")
+
+        # Below variable is being used for the accounting log match
+        # which is currently blocked on PTL bug.
```
Comment (bhroam) [inline test/tests/functional/pbs_node_rampdown.py]: Please name the PTL bug (PP ticket) so when it is fixed, we know to uncomment the code.

**Comment by bhroam** [comment]
Thanks @anamikau, please squash again again for the final merge.

**Comment by anamikau** [comment]
@bhroam and @bayucan, one more push to fix the last codacy error/warning. Hope it is happy now.

**Comment by bhroam** [comment]
Still looks good.  Thanks for making the change to make codacy happy.

**Comment by anamikau** [comment]
@PBSPro/pbspro-maintainers , finally I have all checks passed please take it forward.

**Comment by anamikau** [comment]
rebased again ... waiting for checks to pass.

**Comment by anamikau** [comment]
@PBSPro/pbspro-maintainers , please take it forward. All checks have passed.

**Comment by arungrover** [comment]
This pull has been rebased multiple times, in the interest of time I'm merging it in.
